### PR TITLE
API, history contents: allow filters, limit/offset, and ordering

### PR DESCRIPTION
**Pull request by carlfeberhard**
Related to: #1490, #1374

~~WIP~~

~~(As a heads up) this drops some backwards compatibility:~~
~~\- `types` as a keyword is no longer supported~~
~~\- `ids` is gone. It was problematic since HDCAs b/c they're subject to id collision with the HDAs~~
~~\- `deleted` and `visible` are now accessible via filter instead ('?q=deleted&qv=True')~~

### CODE REVIEW

**Comment by nsoranzo** [comment]
As BioBlend co-maintainer I'm worried about the removals, which will affect `bioblend.galaxy.histories.HistoryClient.show_history()` method. In particular:
- `ids` is not exposed, so no problem
- `types` is available since May 2014, but was not documented, so ok to remove too
- `deleted` and `visible`: can these remain supported?

**Comment by carlfeberhard** [comment]
Cool! Thanks for replying about that. This is the kind of info and feedback I was hoping for with the wip tag.

The deleted and visible behavior can certainly still be achieved:
- `deleted=True` --> `q=deleted&qv=True` (or False) 
- `visible=False` --> `q=visible&qv=False` (or True)
- `deleted=False&visible=True` --> `q=deleted&qv=True&q=visible&qv=False`

It looks like bioblend defaults to None for both, is that correct? In other words, "show all regardless of deleted or visible"? If the behavior is acheivable through a different query string, do you still need the two kwargs? IOW, can you bury my behavior behind your client function and keep your API the same?

File: lib/galaxy/webapps/galaxy/api/history_contents.py [hunk 1]
```diff
@@ -29,85 +30,135 @@ def __init__( self, app ):
...
+        The following are optional parameters:
+            view:   string, one of ('summary','detailed'), defaults to 'summary'
+                    controls which set of properties to return
+            keys:   comma separated strings, unused by default
+                    keys/names of individual properties to return
+
+        If neither keys or views are sent, the default view (set of keys) is returned.
+        If both a view and keys are sent, the key list and the view's keys are
+        combined.
+        If keys are send and no view, only those properties in keys are returned.
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/history_contents.py]: s/send/sent/

**Comment by nsoranzo** [comment]
@carlfeberhard Yes, this should be possible, BUT I'm worried about breaking the user experience of people using a not-updated version of BioBlend to connect to a 16.04 server. My suggestion would be to preserve the behaviour at the Galaxy level, e.g.:

```
filter_params = self.parse_filter_params( kwd )
deleted_kwd = kwd.get( 'deleted', None )
if deleted_kwd is not None:
    filter_params.append('deleted', 'eq', deleted_kwd)
visible_kwd = kwd.get( 'visible', None )
if visible_kwd is not None:
    filter_params.append('visible', 'eq', visible_kwd)
filters = self.history_contents_filters.parse_filters(filter_params)
```

**Comment by jmchilton** [comment]
I'm with @nsoranzo here - I'm not one to say we can never break backward compatibility but that query parameter is easy to implement on top of the new schema. As would restoring `types` which is also exposed in [bioblend](https://github.com/galaxyproject/bioblend/blob/master/bioblend/galaxy/histories/__init__.py#L103).

I'm willing to do the leg work here if you would like me to incorporate these changes and write some tests. We can deprecate the functionality by removing it from the docstring.

**Comment by carlfeberhard** [comment]
@nsoranzo Thanks for the response. I'd respectfully submit that we should think twice before we have an un-updated client be the driving use case for not simplifying the API on Galaxy.

How about this?: I'll re-add the deleted and visible keyword parameters for activating those filters but they're then considered deprecated and would be removed after 16.04 is released. `types` and `ids` would be removed now (thanks for that). Does that work for you?

@jmchilton No - thanks for the offer, though.

**Comment by jmchilton** [comment]
@carlfeberhard I agree we need to think twice, but I'd also argue we would need to think twice about it not being a driving use case as well. We have a team dedicated to usegalaxy.org - others have public servers with considerably less than 1 FTE charged with maintaining Galaxy. The result is the updates don't occur and when they do are painful. As a result, backward compatibility is important - not paramount - but an important factor to consider.

**Comment by carlfeberhard** [comment]
@nsoranzo never heard back from you on the above question.

In any case, I went ahead and found a way around. The outdated bioblend libraries won't have a problem.

**Comment by nsoranzo** [comment]
@carlfeberhard Sorry, I did not reply because John already summarized my view on this. Anyway, thanks a lot for maintaining the backward compatibility!

**Comment by jmchilton** [comment]
Many people on this team work very hard to preserve backward compatibility when it is reasonable on bioblend, on other API endpoints, on tools, on workflows, and the tool shed - it is one thing for one of us to ignore backward compatibility but to actually resist efforts to allow others to do the work feels very unhealthily territorial. I'm disappointed this PR was merged without more discussion.

**Comment by dannon** [comment]
> Many people on this team work very hard to preserve backward compatibility when it is reasonable on bioblend, on other API endpoints, on tools, on workflows, and the tool shed - it is one thing for one of us to ignore backward compatibility but to actually resist efforts to allow others to do the work feels very unhealthily territorial. I'm disappointed this PR was merged without more discussion.

@jmchilton I'm confused -- @carlfeberhard addressed the backwards compatibility concerns.  Was there more you wanted to talk about?

**Comment by martenson** [comment]
@jmchilton I believe the `deleted` and `visible` behaviors are preserved - which were the only concerns @nsoranzo raised. Are there more?

**Comment by dannon** [comment]
ping @jmchilton - never heard back but this is pretty important -- were there other changes that we need to address?

**Comment by martenson** [comment]
@dannon I would consider this solved, John will speak up if he wants to.

**Pull request merged**